src/client/components/filter-menu/time-filter-menu/preset-time-tab.tsx

```diff
@@ -74,16 +75,57 @@ export class PresetTimeTab extends React.Component<PresetTimeTabProps, PresetTim
 
   saveTimeFilter = () => {
     if (!this.validate()) return;
-    const { clicker, onClose, essence, dimension: { name: dimensionName } } = this.props;
-    const { filterPeriod, filterDuration, timeShift } = this.state;
-    clicker.changeFilter(essence.filter.setClause(constructFilter(filterPeriod, filterDuration, dimensionName)));
-    clicker.changeComparisonShift(TimeShift.fromJS(timeShift));
+    const { clicker, onClose } = this.props;
+    clicker.changeFilter(this.constructRelativeFilter());
+    clicker.changeComparisonShift(this.constructTimeShift());
     onClose();
   }
 
+  constructTimeShift(): TimeShift {
+    return TimeShift.fromJS(this.state.timeShift);
+  }
+
+  constructRelativeFilter(): Filter {
+    const { essence, dimension: { name: dimensionName } } = this.props;
+    const { filterPeriod, filterDuration } = this.state;
+    return essence.filter.setClause(constructFilter(filterPeriod, filterDuration, dimensionName));
+  }
+
+  doesTimeShiftOverlap(): boolean {
+    const timeShift = this.constructTimeShift();
+    if (timeShift.isEmpty()) return false;
+    const timeShiftDuration = timeShift.valueOf();
+    const filterDuration = Duration.fromJS(this.state.filterDuration);
+    return filterDuration.getCanonicalLength() > timeShiftDuration.getCanonicalLength();
+  }
+
+  isTimeShiftValid(): boolean {
+    return isTimeShiftValid(this.state.timeShift);
+  }
+
+  isDurationValid(): boolean {
+    return isValidDuration(this.state.filterDuration);
+  }
+
+  validateOverlap(): string | null {
+    const periodOverlaps = this.isTimeShiftValid() && this.isDurationValid() && this.doesTimeShiftOverlap();
+    return periodOverlaps ? STRINGS.overlappingPeriods : null;
+  }
+
+  isFormValid(): boolean {
+    const { filterPeriod } = this.state;
+    return filterPeriod && this.isDurationValid() && this.isTimeShiftValid() && !this.doesTimeShiftOverlap();
+  }
+
+  isFilterDifferent(): boolean {
+    const { essence: { filter, timeShift } } = this.props;
+    const newTimeShift = this.constructTimeShift();
+    const newFilter = this.constructRelativeFilter();
+    return !filter.equals(newFilter) || !timeShift.equals(newTimeShift);
+  }
+
```

Comment:
Some of the added functions look like util function with can be outside react component. Separate file can be better than class methods? Unit tests will be easer to write.

Reply:
Do you think such unit tests would be meaningful? I'd like to write e2e for checking these behaviours. 

And if you extract methods as functions you would need to pass everywhere whole state and props. I don't think readability would increase much.

Reply:
unit tests for some basic time operations are better than e2e

Reply:
We need more meaningful proposals than extract to function and unit test, to have some discussion.

Half of this function is composing predicates. I don't see any point in testing them, because space of possible inputs is much bigger than real scenarios. I want to test if UI responds to user actions. 

If I extract some subset of these methods and test them, that's still not whole story. I need to orchestrate props/state on component and check if correct functions are called. If extracted functions are used only there, what's the point? I split tests, test intermediate results and reduce maintainability.

Point taken, that comparing canonical lengts could b reused in Essence class, so maybe little part could be extracted. But again, here I got text time shift, in Essence I have full object. Still half of this method is glue code.